train/uft/tuning.py

```diff
@@ -0,0 +1,130 @@
+
+    return {
+        "checkpoint_path": config.get("checkpoint_path"),
+        "images_dir": config.get("images_dir"),
+        "save_dir": config.get("save_dir"),
+        "eval_dir": config.get("eval_dir")
+    }
+
+def get_params(trial):
+    # Generate the optimizers.
+    params = {
+            'denoiser_adam_lr': trial.suggest_float('denoiser_adam_lr', 1e-6, 1e-2, log = True), 
+            'discriminator_adam_lr': trial.suggest_float('discriminator_adam_lr', 1e-6, 1e-2, log = True), 
+            'denoiser_loss_b1' : trial.suggest_int('denoiser_loss_b1', 0.1, 0.9),
+            'denoiser_loss_b2' : trial.suggest_float('denoiser_loss_b2',0.1, 0.9),
+            'print_loss_interval': trial.suggest_int('print_loss_interval', 50, 500),
+            'denoiser_adversarial_loss_clip_min': trial.suggest_float('denoiser_adversarial_loss_clip_min', 0.0, 1.0),
+            'denoiser_adversarial_loss_clip_max': trial.suggest_float('denoiser_adversarial_loss_clip_max', 0.0, 1.0),
+            'calc_eval_loss_interval': trial.suggest_int("calc_eval_loss_interval", 50, 500),
+    }
+
+    return params
+
+def objective (trial):
+    config_path = Path(__file__).parent / "config.yml"
+    images_dir =  load_directory_from_yml(config_path)["images_dir"]
+    num_epochs = trial.suggest_int('num_epochs', 1, 2)
+    batch_size = trial.suggest_int('batch_size', 4, 24, step = 4)
+    train_ratio = 0.7
+    checkpoint_path = load_directory_from_yml(config_path)["checkpoint_path"]
+
+    save_dir = load_directory_from_yml(config_path)["save_dir"]
+    save_dir = create_unique_save_dir(save_dir)
+
+    current_run_save_dir = create_dir_with_hyperparam_name(
+            save_dir, get_params(trial)["denoiser_loss_b1"], get_params(trial)["denoiser_loss_b2"]
+    )
+
+    param_save_dir = current_run_save_dir
+
+    train_datasets, val_datasets = create_train_and_validation_unpaired_datasets(
+            images_dir, num_epochs, train_ratio=train_ratio
+        )
+        
+
+    update_key_if_new_value_is_not_none(get_params(trial), "checkpoint_path", checkpoint_path)
+    (
+        _,
+        (denoiser_loss_records, discriminator_loss_records),
+        (
+                val_loss_computed_indices,
+                val_denoiser_loss_records,
+                val_discriminator_loss_records,
+        ),
+    )= train_loop(train_datasets, val_datasets, checkpoint_path, param_save_dir, batch_size, num_epochs, **get_params(trial))
+
+    eval_dir = load_directory_from_yml(config_path)["eval_dir"]
+
+    # evaluation
+    dataset = create_evaluation_dataset(eval_dir)
+    dataloader = DataLoader(dataset, batch_size= batch_size, drop_last=True)
+    psnr_per_sample, ssim_per_sample = evaluate(
+        dataloader,
+        save_dir,
+        checkpoint_path,
+        save_images= "sample"
+    )
+
+    avg_psnr = psnr_per_sample.mean().item()
+    avg_ssim = ssim_per_sample.mean().item()
+    
+    return avg_psnr + avg_ssim
```

Comment:
actually we can't simply add these values because PSNR usally takes a reange of 30~50 while SSIM is always within a range 0 ~ 1. 

two possible solutions
1. normalize the PSNR value within the range of 0 to 1, for fair comparison
2. simply use PSNR only (using SSIM only might cause overfitting because we use SSIM as loss)

i guess we can do the second solution because it's simpler.

Reply:
ok i change to PSNR only